File: json-schema/research/format_hostname/runner.py
#!/usr/bin/env python3
"""Python runner for the `hostname` format conformance corpus.

Implements the PINNED generator-owned check for the Python target:
  1. module-level re.compile(pattern, re.ASCII), fully anchored.
     The end anchor is written `\\Z` (Python's STRICT end-of-string anchor);
     plain `$` in Python matches before a trailing '\\n', which would diverge
     from Go/JS -- exactly the pattern-spec `$`->`\\Z` normalization.
  2. a total-length guard (1..=253 code points) OUTSIDE the regex.
Verdict = (regex matches) AND (length in range).

Emits JSON Lines: {"id","engine":"python","valid","regex","len_ok"}
Run: python3 runner.py [corpus.json]
"""
import json
import re
import sys

HOST_RE = re.compile(
    r"\A[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?"
    r"(?:\.[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)*\Z",
    re.ASCII,
)
MAX_TOTAL_LEN = 253


def main() -> None:
    path = sys.argv[1] if len(sys.argv) > 1 else "corpus.json"
    with open(path, encoding="utf-8") as fh:
        corpus = json.load(fh)
    out = sys.stdout
    for k in corpus["cases"]:
        inst = k["instance"]
        n = len(inst)  # Python str len is code points
        len_ok = 1 <= n <= MAX_TOTAL_LEN
        regex = HOST_RE.search(inst) is not None
        out.write(
            json.dumps(
                {
                    "id": k["id"],
                    "engine": "python",
                    "valid": regex and len_ok,
                    "regex": regex,
                    "len_ok": len_ok,
                },
                ensure_ascii=False,
            )
            + "\n"
        )


if __name__ == "__main__":
    main()

File: specs/json-schema/research/format_hostname/runner.rb
# encoding: UTF-8
# frozen_string_literal: true
#
# Ruby runner for the JSON-Schema `hostname` format conformance corpus.
#
# Implements the PINNED generator-owned check for the (prospective) Ruby target:
#   1. a compiled Regexp. Ruby's `^`/`$` are LINE anchors with no non-multiline
#      mode, so we anchor with \A ... \z (start / strict end of STRING), exactly
#      the pattern-spec Ruby normalization. \d/\w are not used here (the class is
#      an explicit [A-Za-z0-9-]), and there is no \b, so no (?a) flag is needed.
#   2. a total-length guard (1..253 code points) OUTSIDE the regex. Ruby
#      String#length counts code points, so it agrees with the other targets.
#   Verdict = match? AND length-in-range. match? is unanchored search, fine
#   because the pattern is fully \A..\z anchored.
#
# Emits JSON Lines: {"id","engine":"ruby","valid","regex","len_ok"}
# Run: ruby runner.rb [corpus.json]

require "json"

HOST_RE = /\A[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?(?:\.[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)*\z/
MAX_TOTAL_LEN = 253

def main
  path = ARGV[0] || "corpus.json"
  corpus = JSON.parse(File.read(path, encoding: "UTF-8"))
  corpus["cases"].each do |k|
    inst = k["instance"]
    n = inst.length # code points
    len_ok = n >= 1 && n <= MAX_TOTAL_LEN
    regex = HOST_RE.match?(inst)
    $stdout.puts JSON.generate("id" => k["id"], "engine" => "ruby",
                               "valid" => (regex && len_ok),
                               "regex" => regex, "len_ok" => len_ok)
  end
end

main if __FILE__ == $PROGRAM_NAME
